### PR TITLE
Change runtime wiring redefinition checks to be strict by default

### PR DESCRIPTION
**Pull request by dondonz**
## The problem

As I reported in #3760 I was stumped by my code not working in prod where integration tests seemed fine. It turned out that the type wiring that I thought was brand new, was actually redefining another type in a long forgotten part of the codebase. 

This was a particularly tough bug because the order that Spring for GraphQL loaded the type wirings was non-deterministic, so you never knew which type wiring you got in prod. Of course Spring for GraphQL hasn't done anything wrong here, it is hard to distinguish which of two different wirings you want to use.

I think I haven't noticed this as a problem before, because when writing small tests in this repo, the order that datafetchers (and other elements) are redefined/clobbered are predictable.

## The solution

It turns out we already had a "strict mode" option to catch a number of type runtime wiring redefinitions, including type redefinition, as well as redefinitions on datafetchers, default datafetchers, scalars, and enums. However, it's not enabled by default so it's usually not switched on.

For most users, I think having strict checks for redefinition is the better default option. It is very strange to have two type runtime wirings for the same type, and one can clobber the other in an _unpredictable_ way. 

The only time we should allow more than 1 TypeRuntimeWiring definition for the same type is when adding DataFetchers for distinct child fields. By definition, we can only have one type resolver and one enum values provider in a TypeRuntimeWiring, so these must not be redefined.

This PR changes the default to strict checks set to true.

This will be a "breaking change", but it is possible to return to the previous behaviour.

### CODE REVIEW

File: src/test/groovy/graphql/execution/incremental/DeferExecutionSupportIntegrationTest.groovy [hunk 1]
```diff
@@ -162,19 +162,25 @@ class DeferExecutionSupportIntegrationTest extends Specification {
...
-                .type(newTypeWiring("Post").dataFetcher("dataFetcherError", resolveWithException()))
-                .type(newTypeWiring("Post").dataFetcher("dataAndError", resolveWithDataAndError("data")))
-                .type(newTypeWiring("Post").dataFetcher("coercionError", resolve("Not a number", 10)))
-                .type(newTypeWiring("Post").dataFetcher("typeMismatchError", resolve([a: "A Map instead of a List"], 10)))
-                .type(newTypeWiring("Post").dataFetcher("nonNullableError", resolve(null)))
-                .type(newTypeWiring("Page").dataFetcher("summary", resolve("A page summary", 10)))
-                .type(newTypeWiring("Page").dataFetcher("text", resolve("The page full text", 100)))
-                .type(newTypeWiring("Comment").dataFetcher("content", resolve("Full content", 100)))
-                .type(newTypeWiring("Comment").dataFetcher("author", resolve([name: "Author name"], 10)))
+                .type(newTypeWiring("Post")
```
Comment (dondonz) [inline src/test/groovy/graphql/execution/incremental/DeferExecutionSupportIntegrationTest.groovy:165]: Some test updates, extra datafetcher registrations should happen inside the same newTypeWiring builder

File: src/test/groovy/graphql/schema/idl/RuntimeWiringTest.groovy [hunk 2]
```diff
@@ -193,15 +193,14 @@ class RuntimeWiringTest extends Specification {
         newWiring.fieldVisibility == fieldVisibility
     }
 
-    def "strict mode can stop certain redefinitions"() {
+    def "strict mode, on by default, can stop certain redefinitions"() {
```
Comment (dondonz) [inline src/test/groovy/graphql/schema/idl/RuntimeWiringTest.groovy:196]: Updating tests and adding new ones to lock in the change in default behaviour

File: src/main/java/graphql/schema/idl/RuntimeWiring.java [hunk 3]
```diff
@@ -192,11 +192,24 @@ private Builder() {
...
+        }
+
         /**
          * This puts the builder into strict mode, so if things get defined twice, for example, it will throw a {@link StrictModeWiringException}.
          *
          * @return this builder
+         *
+         * @deprecated strictMode default value changed to true, use {@link #strictMode(boolean)} instead
          */
+        @Deprecated(since = "2025-02-15", forRemoval = true)
```
Comment (dondonz) [inline src/main/java/graphql/schema/idl/RuntimeWiring.java:212]: Setting `forRemoval = true` to give clear indication this is going to be removed soon

File: src/main/java/graphql/schema/idl/RuntimeWiring.java [hunk 4]
```diff
@@ -307,6 +320,9 @@ public Builder type(TypeRuntimeWiring typeRuntimeWiring) {
 
             DataFetcher<?> defaultDataFetcher = typeRuntimeWiring.getDefaultDataFetcher();
             if (defaultDataFetcher != null) {
+                if (strictMode && defaultDataFetchers.containsKey(typeName)) {
+                    throw new StrictModeWiringException(format("The type %s already has a default data fetcher defined", typeName));
+                }
```
Comment (dondonz) [inline src/main/java/graphql/schema/idl/RuntimeWiring.java:325]: An edge case that wasn't previously covered: if a type wiring's default datafetcher was redefined at the runtime wiring level (and not the type wiring level), previously didn't raise an error

See test below to lock in the new behaviour

File: src/test/groovy/graphql/execution/ExecutionStepInfoTest.groovy [hunk 5]
```diff
@@ -160,8 +160,10 @@ class ExecutionStepInfoTest extends Specification {
 
         def runtimeWiring = newRuntimeWiring()
                 .type(newTypeWiring("Query").dataFetcher("hero", samwiseDF))
-                .type(newTypeWiring("User").dataFetcher("friends", friendsDF))
-                .type(newTypeWiring("User").dataFetcher("mates", friendsDF))
+                .type(newTypeWiring("User")
+                        .dataFetcher("friends", friendsDF)
+                        .dataFetcher("mates", friendsDF)
+                )
```
Comment (dondonz) [inline src/test/groovy/graphql/execution/ExecutionStepInfoTest.groovy:166]: Set datafetchers so it doesn't count as type redefinition

File: src/test/groovy/graphql/schema/idl/RuntimeWiringTest.groovy [hunk 6]
```diff
@@ -193,15 +193,14 @@ class RuntimeWiringTest extends Specification {
...
         DataFetcher DF1 = env -> "x"
         DataFetcher DF2 = env -> "x"
         TypeResolver TR1 = env -> null
         EnumValuesProvider EVP1 = name -> null
 
         when:
         RuntimeWiring.newRuntimeWiring()
-                .strictMode()
```
Comment (dondonz) [inline src/test/groovy/graphql/schema/idl/RuntimeWiringTest.groovy:204]: Remove these lines, as the default is strictness on

File: src/test/groovy/graphql/schema/idl/RuntimeWiringTest.groovy [hunk 7]
```diff
@@ -248,34 +243,91 @@ class RuntimeWiringTest extends Specification {
         e5.message == "The type Foo has already has a default data fetcher defined"
     }
 
-    def "overwrite default data fetchers if they are null"() {
-
+    def "strict mode, if set to off, won't stop certain redefinitions"() {
```
Comment (dondonz) [inline src/test/groovy/graphql/schema/idl/RuntimeWiringTest.groovy:246]: Lock in the strict mode off case

File: src/test/groovy/graphql/schema/idl/RuntimeWiringTest.groovy [hunk 8]
```diff
@@ -248,34 +243,91 @@ class RuntimeWiringTest extends Specification {
...
+        // Do not permit redefinition of the default datafetcher
+        when:
+        RuntimeWiring.newRuntimeWiring()
+                .type(TypeRuntimeWiring.newTypeWiring("Foo").defaultDataFetcher(DEFAULT_DF))
+                .type(TypeRuntimeWiring.newTypeWiring("Foo").defaultDataFetcher(DEFAULT_DF2))
+                .build()
 
+        then:
+        def error = thrown(StrictModeWiringException)
+        error.message == "The type Foo already has a default data fetcher defined"
```
Comment (dondonz) [inline src/test/groovy/graphql/schema/idl/RuntimeWiringTest.groovy:331]: Covering a new edge case, where the default datafetcher is redefined at the runtime wiring level

We already had a test for when it is redefined inside the type wiring

File: src/test/groovy/graphql/schema/idl/TypeRuntimeWiringTest.groovy [hunk 9]
```diff
@@ -7,42 +7,41 @@ import spock.lang.Specification
 class TypeRuntimeWiringTest extends Specification {
 
     void setup() {
-        TypeRuntimeWiring.setStrictModeJvmWide(false)
+        TypeRuntimeWiring.setStrictModeJvmWide(true)
     }
 
     void cleanup() {
-        TypeRuntimeWiring.setStrictModeJvmWide(false)
+        TypeRuntimeWiring.setStrictModeJvmWide(true)
```
Comment (dondonz) [inline src/test/groovy/graphql/schema/idl/TypeRuntimeWiringTest.groovy:14]: Changes to tests to reflect the new default

File: src/test/groovy/graphql/schema/idl/TypeRuntimeWiringTest.groovy [hunk 10]
```diff
@@ -51,38 +50,36 @@ class TypeRuntimeWiringTest extends Specification {
         e.message == "The field foo already has a data fetcher defined"
     }
 
-    def "strict mode can be turned on JVM wide"() {
-
-
+    def "strict mode can be turned off and on JVM wide"() {
```
Comment (dondonz) [inline src/test/groovy/graphql/schema/idl/TypeRuntimeWiringTest.groovy:53]: Invert the test

**Review by bbakerman — APPROVED** [review]
